### PR TITLE
Add eval_steps as a flag

### PR DESCRIPTION
**Pull request by Infernaught**
Add eval_steps under trainer this time instead of preprocessing

### CODE REVIEW

File: ludwig/models/predictor.py [hunk 1]
```diff
@@ -234,7 +234,12 @@ def batch_evaluation(self, dataset, collect_predictions=False, collect_logits=Fa
                 progress_bar = LudwigProgressBar(self.report_tqdm_to_ray, progress_bar_config, self.is_coordinator())
 
                 predictions = defaultdict(list)
+                eval_steps = self.dist_model.config_obj.trainer.eval_steps
```
Comment (justinxzhao) [inline ludwig/models/predictor.py:237]: Maybe to accommodate GBMs, do:

```python
eval_steps = self.dist_model.config_obj.trainer.eval_steps if hasattr(self.dist_model.config_obj.trainer, "eval_steps") else None
```

File: ludwig/schema/trainer.py [hunk 2]
```diff
@@ -207,6 +207,13 @@ def __post_init__(self):
         parameter_metadata=TRAINER_METADATA[MODEL_ECD]["train_steps"],
     )
 
+    eval_steps: float = schema_utils.NonNegativeInteger(
```
Comment (justinxzhao) [inline ludwig/schema/trainer.py:210]: Another option is to only support `eval_steps` for the LLMTrainer config. We're more concerned about slow synchronous evals for fine-tuning LLMs anyway. WDYT? CC: @geoffreyangus

Comment (Infernaught) [reply ludwig/schema/trainer.py:210]: That's kinda what I was thinking, too

File: ludwig/schema/metadata/configs/trainer.yaml [hunk 3]
```diff
@@ -382,6 +382,22 @@ ecd:
             through the dataset to converge while others converge before a single epoch
             through the data.
         ui_display_name: Train Steps
+    eval_steps:
+        default_value_reasoning:
+            The default value is None because we do not want to lower the number of evaluation steps
+            by default, and we do not know the size of an arbitrary dataset.
+            By setting the default to None, we simply evaluate on the full evaluation set.
+        description_implications:
+            The smaller this value of this parameter, the less time evaluation will take.
```
Comment (geoffreyangus) [inline ludwig/schema/metadata/configs/trainer.yaml:391]: nit: would also mention that smaller values make the evaluation metrics less precise.

File: ludwig/models/predictor.py [hunk 4]
```diff
@@ -366,7 +376,17 @@ def batch_evaluation(self, dataset, collect_predictions=False, collect_logits=Fa
...
+                eval_steps = (
+                    self.dist_model.config_obj.trainer.eval_steps
+                    if hasattr(self.dist_model, "config_obj")
+                    and hasattr(self.dist_model.config_obj.trainer, "eval_steps")
+                    else None
+                )
+                eval_steps_counter = 0
                 while not batcher.last_batch():
+                    if eval_steps and eval_steps_counter >= eval_steps:
+                        logger.info(f"Reached evaluation step {eval_steps}. Ending evaluation.")
```
Comment (geoffreyangus) [inline ludwig/models/predictor.py:388]: @Infernaught can you write a test that confirms the first `n` samples of the evaluation are what are evaluated against per epoch?

**Review by geoffreyangus — COMMENTED** [review]
Nice job so far! The key thing missing is some guarantee that the first `eval_steps` samples of the evaluation step are what are evaluated against. Its probably the case, but we should be sure.

File: ludwig/models/predictor.py (hunk 4 above)
Comment (Infernaught) [reply ludwig/models/predictor.py:388]: Done!

File: tests/ludwig/evaluation/test_evaluation.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    results1 = model.evaluate(df)
+
+    model.config_obj.trainer.eval_steps = 4
+    results2 = model.evaluate(df)
+    results3 = model.evaluate(df)
+
+    diff_exists = False
+    for k in results1[0]["out"]:
+        # Some metrics will be 0 across all runs
+        if results1[0]["out"][k] != 0 and results1[0]["out"][k] != results2[0]["out"][k]:
```
Comment (justinxzhao) [inline tests/ludwig/evaluation/test_evaluation.py:62]: Do we want to check somewhere in this test that there are 4 sample outputs?

**Pull request merged**